crates/ruff_python_formatter/src/verbatim.rs

```diff
@@ -0,0 +1,505 @@
+use std::iter::FusedIterator;
+
+use ruff_formatter::write;
+use ruff_python_ast::node::AnyNodeRef;
+use ruff_python_ast::{Ranged, Stmt};
+use ruff_python_trivia::lines_before;
+use ruff_text_size::TextRange;
+
+use crate::comments::format::{empty_lines, format_comment};
+use crate::comments::{leading_comments, trailing_comments, SourceComment};
+use crate::prelude::*;
+use crate::statement::suite::SuiteChildStatement;
+use crate::verbatim_text;
+
+/// Disables formatting for all statements between the `first_suppressed` that has a leading `fmt: off` comment
+/// and the first trailing or leading `fmt: on` comment. The statements are formatted as they appear in the source code.
+///
+/// Returns the last formatted statement.
+///
+/// ## Panics
+/// If `first_suppressed` has no leading suppression comment.
+#[cold]
+pub(crate) fn write_suppressed_statements_starting_with_leading_comment<'a>(
+    // The first suppressed statement
+    first_suppressed: SuiteChildStatement<'a>,
+    statements: &mut std::slice::Iter<'a, Stmt>,
+    f: &mut PyFormatter,
+) -> FormatResult<&'a Stmt> {
+    let comments = f.context().comments().clone();
+    let source = f.context().source();
+
+    let mut leading_comment_ranges =
+        CommentRangeIter::outside_suppression(comments.leading_comments(first_suppressed), source);
+
+    let before_format_off = leading_comment_ranges
+        .next()
+        .expect("Suppressed node to have leading comments");
+
+    let (formatted_comments, format_off_comment) = before_format_off.unwrap_suppression_starts();
+
+    // Format the leading comments before the fmt off
```

Comment:
Nit: include an example, showing comments before the `fmt off` that we'd want to handle here? (I follow what's going on, but it might be nice to show it visually.)